scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +741,336 @@ class ProcessMonitor(ScalyrMonitor):
-      app.disk.requests type=read:      the number of disk requests.
-      app.disk.bytes type=write:        the number of bytes written to disk
-      app.disk.requests type=write:     the number of disk requests.
-      app.io.fds type=open:             the number of file descriptors held open by the process
 
     In additional to the fields listed above, each metric will also have a field 'app' set to the monitor id to specify
     which process the metric belongs to.
 
     You can run multiple instances of this monitor per agent to monitor different processes.
     """
+
+    # poll interval (in seconds) for matching processes when we are monitoring multiple processes
+    # Needs to be initialized as a class variable, not an instance variable so that it is
+    # alive for all epochs of the monitor.
+    LAST_POLLED = {}
+
     def _initialize(self):
         """Performs monitor-specific initialization."""
-        # The id of the process being monitored, if one has been matched.
-        self.__pid = None
-        # The list of BaseReaders instantiated to gather metrics for the process.
-        self.__gathers = []
+        # The ids of the processes being monitored, if one has been matched.
+        self.__pids = None
 
         self.__id = self._config.get('id', required_field=True, convert_to=str)
         self.__commandline_matcher = self._config.get('commandline', default=None, convert_to=str)
-        self.__target_pid = self._config.get('pid', default=None, convert_to=str)
-
-        if self.__commandline_matcher is None and self.__target_pid is None:
-            raise BadMonitorConfiguration('At least one of the following fields must be provide: commandline or pid',
-                                          'commandline')
+        self.__aggregate_multiple_processes = self._config.get(
+            'aggregate_multiple_processes', default=False, convert_to=bool
+        )
+        self.__include_child_processes = self._config.get(
+            'include_child_processes', default=False, convert_to=bool
+        )
+
+        self.__poll_interval = self._config.get('poll_interval', default=120, convert_to=int)
+        self.__datetime_format = "%Y-%m-%d %H:%M:%S"
+        self.__last_polled_env_key = '%s_linux_process_monitor_last_polled' % self.__id
+
+        __target_pids = self._config.get('pid', default=None, convert_to=str)
+        self.__target_pids = []
+
+        # convert target pid into a list. Target pids can be a single pid or a CSV of pids
+        if __target_pids:
+            for _t in __target_pids.split(','):
+                if _t:
+                    if _t != '$$':
+                        self.__target_pids.append(int(_t))
+                    else:
+                        self.__target_pids.append(int(os.getpid()))
+
+        # history of all metrics which has form:
+        # {
+        #   '<process id>: {
+        #                 <metric name>: [<metric at time 0>, <metric at time 1>.... ],
```

Comment:
This makes it sound like we are keeping more than two samples... I would describe it just as the last two values for all the metrics.